File: P01/Backend/Node/main.js
const express = require('express');
const cors = require('cors');
const { Inventario } = require('./utils/inventario');

const app = express();
app.use(express.json());
// Uso de cors para aceptar peticiones front (se puede ajustar más dependiendo de lo que se requiera)
app.use(cors());

const controlador = new Inventario();

// Endpoint para obtener productos ej: /productos
app.get('/productos', (req, res) => {
    const productos = controlador.listarProductos();
    res.json(productos.map(producto => producto.retornoProducto()));
});

// Endpoint para agregar productos ej: /productos/agregar
/*
{
    "p_nombre": "Pizza",
    "p_cantidad": 12,
    "p_precio": 12.52
}
*/
app.post('/productos/agregar', (req, res) => {
    const { p_nombre, p_cantidad, p_precio } = req.body;
    try {
        controlador.nuevoProducto(p_nombre, p_cantidad, p_precio);
        res.status(201).json({ message: "Producto agregado exitosamente" });
    } catch (error) {
        res.status(400).json({ error: error.message });
    }
});

// Endpoint para eliminar los productos ej: /productos/quitar/Pizza
app.delete('/productos/quitar/:p_nombre', (req, res) => {
    const { p_nombre } = req.params;
    const success = controlador.eliminarProducto(p_nombre);
    if (success) {
        res.json({ message: `Producto '${p_nombre}' eliminado` });
    } else {
        res.status(404).json({ error: "Producto no encontrado (inexistente)" });
    }
});

// Endpoint para buscar los productos ej: /productos/buscar/Pizza
app.get('/productos/buscar/:p_nombre', (req, res) => {
    const { p_nombre } = req.params;
    const producto = controlador.buscarProducto(p_nombre);
    if (producto) {
        res.json(producto.retornoProducto());
    } else {
        res.status(404).json({ error: "Producto no encontrado (inexistente)" });
    }
});

// Endpoint para ordenar los productos ej: /productos/sortear??precio=true&cantidad=true
app.get('/productos/sortear', (req, res) => {
    const { precio, cantidad } = req.query;
    const productosOrdenados = controlador.ordenarProducto(
        precio === 'true',
        cantidad === 'true'
    );
    res.json(productosOrdenados.map(producto => producto.retornoProducto()));
});

// Puerto en el que se conectará con el backend
const PORT = 3200;
app.listen(PORT, () => {
    console.log(`Servidor en puerto '${PORT}'`);
});

File: P01/Backend/Python/main.py
from flask import Flask, jsonify, request
from utils.inventario import Inventario

app = Flask(__name__)
controlador = Inventario()

@app.route('/productos', methods=['GET'])
def list_products():
    return jsonify([producto.retorno_producto() for producto in controlador.productos])

@app.route('/productos/agregar', methods=['POST'])
def nuevo_producto():
    data = request.json
    try:
        controlador.nuevo_producto(data['p_nombre'], data['p_cantidad'], data['p_precio'])
        return jsonify({"message": "Producto agregado exitosamente"}), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 400

@app.route('/productos/quitar/<string:p_nombre>', methods=['DELETE'])
def eliminar_producto(p_nombre):
    success = controlador.eliminar_producto(p_nombre)
    if success:
        return jsonify({"message": f"Producto '{p_nombre}' eliminado"})
    return jsonify({"error": "Producto no encontrado (inexistente)"}), 404

@app.route('/productos/buscar/<string:p_nombre>', methods=['GET'])
def buscar_producto(p_nombre):
    producto = controlador.buscar_producto(p_nombre)
    if producto:
        return jsonify(producto.retorno_producto())
    return jsonify({"error": "Producto no encontrado (inexistente)"}), 404

@app.route('/productos/sortear', methods=['GET'])
def ordenar_producto():
    precio = request.args.get('precio', False, type=bool)
    cantidad = request.args.get('cantidad', False, type=bool)
    productos_ordenados = controlador.ordenar_producto(precio=precio, cantidad=cantidad)
    return jsonify([producto.retorno_producto() for producto in productos_ordenados])

if __name__ == "__main__":
    app.run(debug=True, port=3200)